upstairs/src/volume.rs

```diff
@@ -0,0 +1,1228 @@
+
+                    if sub_owned_vec[start_of_block] {
+                        // sub volume has written to this block, use it
+                        for block_offset in 0..self.block_size {
+                            let inside_block =
+                                start_of_block + block_offset as usize;
+                            data_vec[data_index + inside_block] =
+                                sub_data_vec[inside_block];
+                        }
+                    } else {
+                        // sub volume hasn't written to this block, use the
+                        // parent
+                        for block_offset in 0..self.block_size {
+                            let inside_block =
+                                start_of_block + block_offset as usize;
+                            data_vec[data_index + inside_block] =
+                                parent_data_vec[inside_block];
+                        }
+                    }
+                }
+            } else {
+                let mut data_vec = data.as_vec();
+
+                // no parent
+                data_vec[data_index..(data_index + sz)]
+                    .copy_from_slice(&sub_buffer.as_vec());
+            }
+
+            data_index += sz;
+        }
+
+        assert_eq!(data.len(), data_index);
+
+        BlockReqWaiter::immediate()
+    }
+
+    fn write(
+        &self,
+        offset: Block,
+        data: Bytes,
+    ) -> Result<BlockReqWaiter, CrucibleError> {
+        // In the cast that this volume only has a read only parent, return an
+        // error.
+        if self.sub_volumes.is_empty() {
+            crucible_bail!(CannotReceiveBlocks, "No sub volumes!");
+        }
+
+        let affected_sub_volumes = self.sub_volumes_for_lba_range(
+            offset.value,
+            data.len() as u64 / self.block_size,
+        );
+
+        // TODO parallel dispatch!
+        let mut data_index = 0;
+        for (coverage, sub_volume) in affected_sub_volumes {
+            let sub_offset = Block::new(
+                sub_volume.compute_sub_volume_lba(coverage.start),
+                offset.shift,
+            );
+            let sz = (coverage.end - coverage.start) as usize
+                * self.block_size as usize;
+            let slice_range = Range::<usize> {
+                start: data_index,
+                end: data_index + sz,
+            };
+            let slice = data.slice(slice_range);
+
+            let mut waiter = sub_volume.write(sub_offset, slice.clone())?;
+            waiter.block_wait()?;
+
+            data_index += sz;
+        }
+
+        BlockReqWaiter::immediate()
+    }
+
+    fn flush(&self) -> Result<BlockReqWaiter, CrucibleError> {
+        for sub_volume in &self.sub_volumes {
+            let mut waiter = sub_volume.flush()?;
+            waiter.block_wait()?;
+        }
+
+        // XXX no need to flush read only parent?
```

Comment:
I hope no flush is needed for read only.
However, we do have to make sure that "read only volumes" have all three downstairs consistent.  Not really a topic for this PR though.

Reply:
Removed this XXX in 80c3b25.